neo/SmartContract/ApplicationEngine.cs

```diff
@@ -604,30 +604,36 @@ private bool PreStepInto(OpCode nextOpcode)
             return true;
         }
 
-        public static ApplicationEngine Run(byte[] script, IScriptContainer container = null, Block persisting_block = null, bool testMode = false)
+        public static ApplicationEngine Run(byte[] script, Snapshot snapshot,
```

Comment:
Why modify this line?

Reply:
The diff is just showing it strangely; this is an additional overloaded method for Run that allows passing a snapshot. The overview of my pull requests explains why I added it. See line 632, it still has the previous method signature of the Run method supported.

Reply:
I don't see why we need to add an overload to allow a Snapshot to be passed in.

Reply:
I can tell you why I need it. By the time I get the message for PersistedBlock, the chain may already have persisted another block in some cases. I need to run some scripts against the block in the chain for which the PersistBlock is being processed. I have made one other modification to NEO core (not in this PR) so that I have the snapshot from the previous block. The other change to provide the snapshot from the block when sending PersistedBlock messages is a bit tricky because it is not really immutable and must be destroyed by exactly one consumer. So I haven’t put a PR for that; however there are also other cases where someone may want to get a snapshot before running scripts and then run a bunch of scripts against that snapshot to ensure they are run in with consistent data from same block (in the case the chain accepts another block in the middle).